File: test/clj/com/rpl/cluster_retriever_test.clj
(ns com.rpl.cluster-retriever-test
  (:use [clojure.test]
        [com.rpl.test-helpers]
        [com.rpl.rama]
        [com.rpl.rama.path])
  (:require
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama.aggs :as aggs]
   [com.rpl.rama.test :as rtest])
  (:import
   [com.rpl.agentorama
    AgentNode]
   [com.rpl.rama.cluster
    ClusterManagerBase]
   [com.rpl.rama
    Depot
    PState
    QueryTopologyClient]))

(deftest cluster-retriever-test
  (with-open [ipc (rtest/create-ipc)]
    (letlocals
     (bind other-module
           (module [setup topologies]
                   (declare-depot setup *depot (hash-by identity))
                   (let [s (stream-topology topologies "s")]
                     (declare-pstate s $$p {clojure.lang.Keyword Long})
                     (<<sources s
                                (source> *depot :> *k)
                                (+compound $$p {*k (aggs/+count)})))
                   (<<query-topology topologies
                                     "q"
                                     [*v :> *res]
                                     (|origin)
                                     (str *v "!" :> *res))))
     (bind other-module-name (get-module-name other-module))

     (bind agent-module
           (module [setup topologies]
                   (let [topology (aor/agent-topology setup topologies)]
                     (->
                      topology
                      (aor/new-agent "foo")
                      (aor/node
                       "start"
                       nil
                       (fn [agent-node k]
                         (let [^ClusterManagerBase retriever (.getClusterRetriever
                                                              ^AgentNode agent-node)
                               ^Depot depot (.clusterDepot retriever
                                                           other-module-name
                                                           "*depot")
                               ^PState p    (.clusterPState retriever
                                                            other-module-name
                                                            "$$p")
                               ^QueryTopologyClient q (.clusterQuery retriever
                                                                     other-module-name
                                                                     "q")
                               res (volatile! [])]
                           (foreign-append! depot k)
                           (vswap! res conj (foreign-select-one (keypath k) p))
                           (vswap! res conj (foreign-select-one (keypath k) p {:pkey k}))
                           (vswap! res conj (foreign-select (keypath k) p))
                           (vswap! res conj (foreign-select (keypath k) p {:pkey k}))
                           (vswap! res conj (foreign-invoke-query q "."))
                           (aor/result! agent-node @res)))))
                     (aor/define-agents! topology))))
     (bind agent-module-name (get-module-name agent-module))

     (rtest/launch-module! ipc other-module {:tasks 1 :threads 1})
     (launch-module-without-eval-agent! ipc agent-module {:tasks 2 :threads 2})

     (bind agent-manager (aor/agent-manager ipc agent-module-name))
     (bind foo (aor/agent-client agent-manager "foo"))

     (bind {:keys [task-id agent-invoke-id] :as inv} (aor/agent-initiate foo :a))
     (bind res (aor/agent-result foo inv))

     (is (= [1 1 [1] [1] ".!"] res)))))

File: test/java/com/rpl/aortest/ClusterRetrieverTest.java
package com.rpl.aortest;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentModule;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentTopology;
import com.rpl.rama.Depot;
import com.rpl.rama.Path;
import com.rpl.rama.PState;
import com.rpl.rama.RamaModule;
import com.rpl.rama.cluster.ClusterManagerBase;
import com.rpl.rama.module.StreamTopology;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Tests for AgentNode.getClusterRetriever. */
public class ClusterRetrieverTest {

  public static class OtherModule implements RamaModule {

    @Override
    public void define(Setup setup, Topologies topologies) {
      setup.declareDepot("*depot", Depot.hashBy("identity"));

      StreamTopology st = topologies.stream("s");
      st.pstate("$$p", PState.mapSchema(String.class, Long.class));
      st.source("*depot")
          .out("*k")
          .localTransform(
              "$$p",
              Path.key("*k").term(c -> c == null ? 1L : (Long) c + 1L));
    }
  }

  private static AgentModule agentModule(String otherModuleName) {
    return new AgentModule() {
      @Override
      protected void defineAgents(AgentTopology topology) {
        topology
            .newAgent("foo")
            .node(
                "start",
                null,
                (AgentNode agentNode, String k) -> {
                  ClusterManagerBase retriever = agentNode.getClusterRetriever();
                  Depot depot = retriever.clusterDepot(otherModuleName, "*depot");
                  PState p = retriever.clusterPState(otherModuleName, "$$p");

                  depot.append(k);

                  Path kpath = Path.key(k);
                  List<Object> res = new ArrayList<>();
                  res.add(p.selectOne(kpath));
                  res.add(p.selectOne(k, kpath));
                  res.add(p.select(kpath));
                  res.add(p.select(k, kpath));

                  agentNode.result(res);
                });
      }
    };
  }

  public static void testClusterRetriever() throws Exception {
    try (InProcessCluster ipc = InProcessCluster.create()) {
      OtherModule otherModule = new OtherModule();
      ipc.launchModule(otherModule, new LaunchConfig(1, 1));

      String otherModuleName = otherModule.getModuleName();
      AgentModule agentModule = agentModule(otherModuleName);
      ipc.launchModule(agentModule, new LaunchConfig(2, 2));

      AgentManager manager = AgentManager.create(ipc, agentModule.getModuleName());
      AgentClient foo = manager.getAgentClient("foo");

      @SuppressWarnings("unchecked")
      List<Object> res = (List<Object>) foo.invoke("a");
      if (res == null) {
        throw new AssertionError("Agent should return a result");
      }
      if (res.size() != 4) {
        throw new AssertionError("Expected 4 results but got " + res.size());
      }
      if (!Long.valueOf(1L).equals(res.get(0))) {
        throw new AssertionError("Expected selectOne result 1 but got " + res.get(0));
      }
      if (!Long.valueOf(1L).equals(res.get(1))) {
        throw new AssertionError("Expected partitioned selectOne result 1 but got " + res.get(1));
      }
      @SuppressWarnings("unchecked")
      List<Long> selectRes = (List<Long>) res.get(2);
      if (!Arrays.asList(1L).equals(selectRes)) {
        throw new AssertionError("Expected select result [1] but got " + selectRes);
      }
      @SuppressWarnings("unchecked")
      List<Long> pkeySelectRes = (List<Long>) res.get(3);
      if (!Arrays.asList(1L).equals(pkeySelectRes)) {
        throw new AssertionError("Expected partitioned select result [1] but got " + pkeySelectRes);
      }
    }
  }

  public static boolean runAllTests() throws Exception {
    System.out.println("Running ClusterRetriever tests...");
    testClusterRetriever();
    System.out.println("✓ testClusterRetriever passed");
    System.out.println("All ClusterRetriever tests passed!");
    return true;
  }
}
